### PR TITLE
Add a search field to the creative inventory

### PR DESCRIPTION
**Pull request by kilbith**
You'll no longer waste your time browsing several pages to find something.
Obviously gives a significant help to the creative activity.

screenshot_20151129_222037

screenshot_20151129_221146

### CODE REVIEW

**Comment by Jordach** [comment]
:+1: This makes sense - why should I (depending on the set of mods), sort
through nearly 30 pages for one block or item? I would like to see tabs -
eg, tools, blocks, misc. This would help in searching faster.

File: mods/creative/init.lua [hunk 1]
```diff
@@ -35,19 +40,29 @@ minetest.after(0, function()
...
-	for name,def in pairs(minetest.registered_items) do
-		if (not def.groups.not_in_creative_inventory or def.groups.not_in_creative_inventory == 0)
-				and def.description and def.description ~= "" then
-			creative_list[#creative_list+1] = name
+	local inv = minetest.get_inventory({type = "detached", name = "creative_" .. player_name})
+
+	for name, def in pairs(minetest.registered_items) do
+		if not (def.groups.not_in_creative_inventory == 1) and
+				def.description and def.description ~= "" then
+			if (filter and def.name:find(filter)) or not filter then
```
Comment (kaeza) [inline mods/creative/init.lua]: This `find` call searches for regular expressions instead of plain substrings. If this is intended, you have to wrap the `find` call in `pcall` and fail gracefully in case of errors. If not intended, change to `find(filter, 1, true)`.

File: mods/creative/init.lua [hunk 2]
```diff
@@ -152,18 +185,18 @@ if minetest.setting_getbool("creative_mode") then
...
-	
+
 	function minetest.handle_node_drops(pos, drops, digger)
 		if not digger or not digger:is_player() then
 			return
 		end
 		local inv = digger:get_inventory()
 		if inv then
-			for _,item in ipairs(drops) do
+			for _, item in pairs(drops) do
```
Comment (kaeza) [inline mods/creative/init.lua]: Why the change to `pairs`?

Comment (kilbith) [inline mods/creative/init.lua]: `ipairs` is completely useless when you don't (need to) iterate over tables with sequential numeric indices. It is the case here.

File: mods/creative/init.lua (hunk 1 above)
Comment (kilbith) [inline mods/creative/init.lua]: It is intended, yes. Sometimes I don't bother to type the whole name to find something, this is more convenient IMO.

Also how it can have errors ? There's always a filtering even with one letter.

Comment (kaeza) [inline mods/creative/init.lua]:
```
# lua
LuaJIT 2.0.2 -- Copyright (C) 2005-2013 Mike Pall. http://luajit.org/
JIT: ON CMOV SSE2 SSE3 fold cse dce fwd dse narrow loop abc sink fuse
> print(("s"):find("%"))
stdin:1: malformed pattern (ends with '%')
stack traceback:
        [C]: in function 'find'
        stdin:1: in main chunk
        [C]: at 0x00401ec0
> print(("s"):find("%", 1, true))
nil
> print(("s"):find("s", 1, true))
1       1
>
```

File: mods/creative/init.lua (hunk 2 above)
Comment (kaeza) [inline mods/creative/init.lua]: `ipairs` makes it clear to the programmer that the table in an array.

Comment (kilbith) [inline mods/creative/init.lua]: You're nitpicking here... don't use knowingly something slower when you don't need to.

File: mods/creative/init.lua (hunk 1 above)
Comment (kilbith) [inline mods/creative/init.lua]: You're right, will change that.

File: mods/creative/init.lua (hunk 2 above)
Comment (kaeza) [inline mods/creative/init.lua]: >slower

```
> iter,c,n,t = pairs,0,5,{"a","b","c"} for it=1, n do s=os.clock() for x = 1, 100000 do for a, b in iter(t) do end end c=c+os.clock()-s end print("Average:", c/n)
Average:        0.020999999999992
> iter,c,n,t = ipairs,0,5,{"a","b","c"} for it=1, n do s=os.clock() for x = 1, 100000 do for a, b in iter(t) do end end c=c+os.clock()-s end print("Average:", c/n)
Average:        0.0032000000000039
```
Still don't see a reason for the change, but won't argue further.

**Comment by C1ffisme** [comment]
This is especially useful on laggy servers, I can't wait to see this (hopefully) merged.

**Comment by JBBgameich** [comment]
I like Jordans Idea of tabs! But i think we need a search too. I also would like to see a better way of getting the items from the crative inventory.  An idea would be if you leftclick on an item, you get 1 of it, and when rightclick on an item you get a full stack. 
-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by C1ffisme** [comment]
One problem about search bars, though (I'm not sure if you can do anything about this though. :( ) : Every time I use one, I'm used to pressing the enter key to search. But this closes my formspec, and I have to type again. Not a problem for me, but to people who point their finger straight downwards and poke each key individually to type, this can get annoying.

**Comment by Jordach** [comment]
@C1ffisme It appears the Unified Inventory doesn't have this issue. Perhaps
fiddling with text inputs might work.

**Comment by kilbith** [comment]
Kinda surprised that none of the maintainers reacted yet...

@PilzAdam @ShadowNinja @Ekdohibs @sfan5  -> a review please ?

**Comment by PilzAdam** [comment]
+1 for the feature, but I currently have no time to look at the code.

File: mods/creative/init.lua (hunk 2 above)
Comment (HybridDog) [inline mods/creative/init.lua]: >Still don't see a reason for the change, but won't argue further.

it takes 556.25% more time

**Comment by HybridDog** [comment]
>[the enter key] closes my formspec

maybe you only need to change the order of the formspec elements,
What happens if you press tab and then enter?

File: mods/creative/init.lua (hunk 2 above)
Comment (kaeza) [inline mods/creative/init.lua]: So in my case, `ipairs` is 5.5 times faster than `pairs`, and in HybridDog's case, `ipairs` is 3.8 times faster. My benchmark is also with LuaJIT, BTW.

But let's change it anyway because...reasons! (not specifically directed to you HybridDog)

**Comment by Ekdohibs** [comment]
:+1: for the feature, code looks good but I haven't reviewed it in detail.

File: mods/creative/init.lua (hunk 2 above)
Comment (kilbith) [inline mods/creative/init.lua]: Changed it back to interrupt the bikeshedding.

**Comment by 0-afflatus** [comment]
:+1: subject to code review; it's a no-brainer really.

**Comment by PilzAdam** [comment]
The problems with the enter key that @C1ffisme pointed out are limitations of the engine. I'm adding the 'need-engine-change' label as a reference, but it shouldn't block this PR.

**Comment by ShadowNinja** [comment]
Engine change: minetest/minetest#516

**Comment by kilbith** [comment]
@PilzAdam O.K. for merging then ? Please do, otherwise nobody else will bother to (not even for reviewing).

**Comment by kaeza** [comment]
>otherwise nobody else will bother to (not even for reviewing).

Offtopic for this discussion, but the issues/PRs keep piling up in this repo with basically nobody ("collaborators" in Github speak) saying yay or nay or even bother to review code, much less to bother merging in the cases where there's a clear agreement. Maybe there's a need to cycle game maintainers?

**Comment by HybridDog** [comment]
>Maybe there's a need to cycle game maintainers?

Actually the repo's master branch is used for the dev version of minetest, isn't it?
So how about adding secondary maintainers merging PRs and if they are unwanted by enough primary ones, the commits get reverted?

**Comment by sofar** [comment]
Reviewed the code, tested it, and seems fine. It creates a proper detached inventory.

I don't see why "needs engine change" is needed at all, it is certainly not a blocking requirement for this patch, and IMO should *not* prevent this from getting merged. The change works, is usable and functions exactly like unified_inventory's search field, so I doubt that anyone would be confused to begin with.

:+1: from me

**Comment by paramat** [comment]
Seems to have enough approval and review, will merge soon (tomorrow).

**Comment by kilbith** [comment]
Thanks you @paramat and @sofar. I appreciate your consideration.

**Comment by paramat** [comment]
b1e2ece63842ff5228798feb9259398c75524f8b
